### PR TITLE
Add a parallel async event handler to OnionMessenger and pass it directly to BackgroundProcessor

### PR DESCRIPTION
**Pull request by TheBlueMatt**
This adds an `OnionMessenger::process_pending_events_async`
mirroring the same in `ChannelManager`. However, unlike the one in
`ChannelManager`, this processes the events in parallel by spawning
all futures and using the new `MultiFuturePoller`.

Because `OnionMessenger` just generates a stream of messages to
store/fetch, we first process all the events to store new messages,
`await` them, then process all the events to fetch stored messages,
ensuring reordering shouldn't result in lost messages (unless we
race with a peer disconnection, which could happen anyway).

### CODE REVIEW

**Comment by valentinewallace** [comment]
Looks like 1.63 builds are sad

File: lightning/src/onion_message/messenger.rs
Comment (valentinewallace) [inline lightning/src/onion_message/messenger.rs:1179]: Could this parallelism affect how users structure their database storage for intercepted OMs? Just wondering if some extra docs could be helpful.

File: lightning/src/onion_message/messenger.rs
Comment (valentinewallace) [reply lightning/src/onion_message/messenger.rs:81]: `AChannelManager`'s docs have the same sentence structure and use periods. Feel free to ignore, though.

**Review by valentinewallace — COMMENTED** [review]
LGTM pending a second reviewer.

File: lightning-background-processor/src/lib.rs
Comment (jkczyz) [inline lightning-background-processor/src/lib.rs:697]: Unclear to me why this is now needed.

**Review by jkczyz — DISMISSED** [review]
LGTM

File: lightning-background-processor/src/lib.rs
Comment (jkczyz) [reply lightning-background-processor/src/lib.rs]: No need for the comment then?

**Comment by valentinewallace** [comment]
LGTM after the comment about `Box` is removed.

**Review by valentinewallace — DISMISSED** [review]

**Review by jkczyz — APPROVED** [review]

**Review by jkczyz — APPROVED** [review]

**Pull request merged**
